### PR TITLE
fix: do not kill process buffer if it was already killed

### LINKED ISSUE #132: `async-start-process`: don't use `(current-buffer)` when killing the process buffer
When the `async-start-process` callback function accidentally kills the processes buffer itself, this kills one other (previous) buffer:

```
(defun async-when-done (proc &optional _change)
  "Process sentinel used to retrieve the value from the child process."
  (when (eq 'exit (process-status proc))
    (with-current-buffer (process-buffer proc)
      (let ((async-current-process proc))
        (if (= 0 (process-exit-status proc))
            (if async-callback-for-process
                (if async-callback
                    (prog1
                        (funcall async-callback proc)
                      (unless async-debug
                        (kill-buffer (current-buffer)))) ;; FIXME: this may erronously kill the wrong buffer
                  (set (make-local-variable 'async-callback-value) proc)
                  (set (make-local-variable 'async-callback-value-set) t))
              (goto-char (point-max))
              (backward-sexp)
              (async-handle-result async-callback (read (current-buffer))
                                   (current-buffer)))
          (set (make-local-variable 'async-callback-value)
               (list 'error
                     (format "Async process '%s' failed with exit code %d"
                             (process-name proc) (process-exit-status proc))))
          (set (make-local-variable 'async-callback-value-set) t))))))
```

I believe the fix is to kill `(process-buffer proc)` instead. You probably should also check if it is still existing. :boom: 

Please also add to the documentation that user callbacks don't need to kill the process's buffer.

### PR DESCRIPTION
**Pull request by Fuco1**
Fixes #132

### CODE REVIEW

File: async.el [hunk 1]
```diff
@@ -171,7 +171,10 @@ It is intended to be used as follows:
                     (prog1
                         (funcall async-callback proc)
                       (unless async-debug
-                        (kill-buffer (current-buffer))))
+                        ;; we need to check this because theoretically
+                        ;; `async-callback' could've killed it already
+                        (when (buffer-live-p (process-buffer proc))
+                          (kill-buffer (process-buffer proc)))))
```
Comment (basil-conto) [inline async.el:177]: It's a safe no-op to pass killed buffers to `kill-buffer`.

OTOH if `process-buffer` returns `nil`, then `kill-buffer` would try to kill the current buffer instead, and that would be a problem.

So perhaps the comment could be clarified? Thanks.

**Comment by thierryvolpiatto** [comment]
Merged, thanks!

File: async.el (hunk 1 above)
Comment (Fuco1) [reply async.el:177]: Yea, previously it was just using `(current-buffer)` which could also point really anywhere if the callback changes buffers for example.

So you're saying that the check is useless entirely?  Or we only need to check if `(process-buffer proc)` itself is not nil?  If it's a buffer, killed or not, kill it, but if no buffer is associated, don't call kill-buffer?.  IDK.  The code as is now kind of makes it clear what's the logic.  

How would you improve the comment, or remove it entirely?

Comment (basil-conto) [reply async.el:177]: > How would you improve the comment, or remove it entirely?

The current comment (and the title of this PR) implies that a check is needed before killing a killed buffer, when what the check is actually doing is checking that `proc` still has a buffer at all.

But that's a pedantic reading, so perhaps the current wording is clear enough.

**Pull request merged**